### DISCUSSION TITLE
How to make a prompt so that its length is proportional to terminal width and pad the prompt with dots if the length of the prompt is less than n

### DISCUSSION THREAD

**Discussion by blueray453** [fish-shell/fish-shell#9823] [Q&A] [answered]
I am thinking of transitioning to fish,  For this I need the prompt which I am currently using with zsh. Here is the details:

```
export PROMPT_PERCENT_OF_LINE=35
width_part='$((${COLUMNS} * ${PROMPT_PERCENT_OF_LINE} / 100))'
 if (($+TMUX)); then
        PS1="%F{013}%$width_part<...< <<\${(l[$width_part][.])}%d%f%F{006}($(tmux display-message -p '#{client_session}(#{session_windows})#{window_index}-#{window_name}'))%f%F{012}$(__git_ps1 "(%s)")%f%F{011}%(1j.jobs%j.)%f%(?.%F{004}.%F{001}✕%?)%<<%# %f"
    else
        export PS1="%F{013}%$width_part<...< <<\${(l[$width_part][.])}%d%f%F{012}$(__git_ps1 "(%s)")%f%F{011}%(1j.jobs%j.)%f%(?.%F{004}.%F{001}✕%?)%<<%# %f"
fi
```

Here, __git_ps1 is taken from by https://github.com/git/git/blob/master/contrib/completion/git-prompt.sh

It looks like:

![prompt](https://github.com/fish-shell/fish-shell/assets/855297/3fab4394-e0e8-4765-9d33-572c6419170d)

More about this prompt can be found on https://unix.stackexchange.com/a/604212/206574 and https://unix.stackexchange.com/a/370276/206574

**Comment by faho** [comment, marked as answer]
Okay, I'm not going to reverse engineer your sigil soup - if you want hieroglyphs translated ask Dr Jones, next floor down.

Instead, here's what you need:

https://fishshell.com/docs/current/cmds/string.html - which has a `string pad` command that can pad a string to a given width.

https://fishshell.com/docs/current/cmds/math.html - to do calculations

https://fishshell.com/docs/current/cmds/fish_git_prompt.html - that's our "git_ps1"

Put it all together in a fish_prompt function, something like

```fish
function fish_prompt
    # Get the actual prompt together - here we use $(prompt_pwd),
    # which shortens it with ~ instead of $HOME and only one char per component.
    # We could also just use $PWD
    # fish_git_prompt does the git display
    set -l prompt "$(fish_git_prompt)$(prompt_pwd)%"
    # Tell string pad to pad it out to the given width with dots
    # (it is smart enough to remove any color escapes from the calculations,
    #  like what fish_git_prompt will print)
    # math's `-s0` makes it truncate to int
    string pad -c . -w (math -s0 "$COLUMNS * 0.35") $prompt
end
```

**Comment by blueray453** [comment]
I checked your code. Algorithmically i think this is what I need. But when running, I am getting an error.

```
% fish
~/.config/fish/functions/fish_prompt.fish (line 6): $(...) is not supported. In fish, please use '(fish_git_prompt)'.
    set -l prompt "$(fish_git_prompt)$(prompt_pwd)%"
                   ^
from sourcing file ~/.config/fish/functions/fish_prompt.fish
in command substitution
source: Error while reading file “/home/ismail/.config/fish/functions/fish_prompt.fish”
Welcome to fish, the friendly interactive shell
Type `help` for instructions on how to use fish
ismail@ismail-zorin /home/ismail >
```

**Reply by faho** [reply]
Your fish version is outdated. See https://fishshell.com/ for how to get up-to-date packages.

If this is Zorin, I believe you need the Ubuntu packages.

**Comment by blueray453** [comment]
I am using Zorin. It says:

```
% sudo apt-add-repository ppa:fish-shell/release-3
LSB codename: 'focal'.
Version of base-files: '11ubuntu5.6+zorin1'.
Your LSB codename isn't a valid Linux Mint codename.
Please check your LSB information with "lsb_release -a".
```

**Reply by blueray453** [reply]
I have a mint vm sitting here. I have tested and it works (except the colors).

![fish](https://github.com/fish-shell/fish-shell/assets/855297/531f4dff-fd15-4afe-9153-82be0dcb6ad5)

Please let me know how to get the colors (%F{013}...%f).

**Reply by faho** [reply]
Your system is misconfigured.

From some cursory googling, it's possible that `sudo apt-get install lsb-core` may help, otherwise you'll have to ask the Zorin people.

Or, alternatively, you can just add the repository to your apt configuration directly: Write

```
deb https://ppa.launchpadcontent.net/fish-shell/release-3/ubuntu focal main
```

into a file in /etc/apt/sources.list.d with the ending ".list".

**Reply by faho** [reply]
> Please let me know how to get the colors (%F{013}...%f).

Fish uses the `set_color` builtin to print color escape sequences. It doesn't use palette colors ("color number 13"), you can either give it named colors like "red" or "brblue" (see `set_color --print-colors`), or hexadecimal rgb like "4512bb":

```fish
set -l prompt "$(fish_git_prompt)$(set_color E6E6FA)$(prompt_pwd)%$(set_color normal)"
```

and fish will then translate that to the appropriate escape sequence - ideally it'll just pass the rgb value to your terminal directly.

**Reply by blueray453** [reply]
Just to be clear, suppose i want to set E6E6FA for fish_git_prompt and FFFFFF for prompt_pwd. I mean set different color for multiple part of the string. How can I do that?

**Reply by faho** [reply]
Just like above: You use multiple [set_color](https://fishshell.com/docs/current/cmds/set_color.html) calls.

```fish
echo $(set_color green)grass$(set_color blue)sky
```

colors grass green and sky blue. `set_color` really just prints an escape sequence that tells the terminal "okay, from here on color the text yellow".

Of course the git prompt can already apply color, depending on its configuration (but I believe that's turned off by default), so you would have to tell it the color.

**Comment by blueray453** [comment]
I installed Latest version of zsh using:

```
sudo sh -c 'echo "deb https://ppa.launchpadcontent.net/fish-shell/release-3/ubuntu focal main" > /etc/apt/sources.list.d/fish-shell.list'
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-key 27557F056DC33CA5
sudo apt update
sudo apt install fish
```

Few more issues with the prompt.

For Tmux The Prompt Should Look Like:
![tmux](https://github.com/fish-shell/fish-shell/assets/855297/f8ed315c-9f24-47d8-9734-1fbc8e827998)

For Background Job The Prompt Should Look Like:
![bg](https://github.com/fish-shell/fish-shell/assets/855297/a4a7f81a-5385-45ed-b704-3791efe16f73)

For Multi-Line Commands, The Second line **should not have wrapping**. 

See 

```
multi line
com
```
position in both shells.

And also **for a non-zero return code, it should print the status in the next line**. Shown Bellow.

![multi-line](https://github.com/fish-shell/fish-shell/assets/855297/41b48aa6-ea58-471a-9996-2918e9c282b7)

There is one more issue.

I activate venv automatically using the following zsh function.

```
function automatically_activate_python_env() {
  if [[ -z $VIRTUAL_ENV ]] ; then
    activate_venv
  else
    parentdir="$(dirname ${VIRTUAL_ENV})"
    if [[ "$PWD"/ != "$parentdir"/* ]] ; then
      deactivate
      activate_venv
    fi
  fi
}

function activate_venv() {
  local d n
  d=$PWD

  until false
  do

  if [[ -f $d/.venv/bin/activate ]] ; then
    source $d/.venv/bin/activate
    break
  fi
    d=${d%/*}
    # d="$(dirname "$d")"
    [[ $d = *\/* ]] || break
  done
}

automatically_activate_python_env "$@"
```

In my .zshrc there is a line like:

    add-zsh-hook precmd automatically_activate_python_venv

How can I do it in fish.

Please help me with these. After that there is one more issue (regarding keybinding and creating plugins and new keybinding functions), then I am all set, hopefully will not bother you any more in the near future.

**Reply by faho** [reply]
I suggest you take a look at the [sample prompts](https://github.com/fish-shell/fish-shell/tree/master/share/tools/web_config/sample_prompts), they do a lot of this stuff.

> I activate venv automatically

Googling "How to activate venv automatically fish" gives me https://github.com/nakulj/auto-venv/blob/main/conf.d/venv.fish, which seems okay.

The gist is you make a function that is triggered `--on-variable PWD` (because of this it needs to be defined eagerly in config.fish or conf.d/, autoloading doesn't work because fish won't know of the trigger before it's run), and then source `activate.fish` (python ships a fish version of the activate script).

> using the following zsh function.

Just as a sidenote: Many of us aren't very familiar with zsh, and so a big function like that isn't very helpful. Just say what you want to accomplish instead of telling us what you did in zsh.

That goes in general - it's easier to learn fish if you try to get away from framing everything in zsh terms. That'll also help you find out if you even *need* a specific bit of configuration.

E.g. I have the feeling that some of your keybindings and such will just already be available by default, so instead of asking "how do I rewrite this function to use xclip so I can bind it to ctrl+c", ask yourself "how do I copy by pressing a key", and then you'll find it's *already* bound to ctrl+x, and already supports xclip and xsel and various wayland things. (but it's not ctrl+c because that's cancelling). Or instead of writing your own history search thing and binding it to ctrl+r, maybe just try out fish's stock history pager first?

**Reply by faho** [reply]
Oh, also:

>For Multi-Line Commands, The Second line should not have wrapping.

The second line of the commandline is drawn at the same column as the first, this isn't configurable.

We don't have a $PS2 or equivalent.

**Reply by blueray453** [reply]
I found that you are doing a rewrite (in rust), if possible then please add this feature.

**Reply by blueray453** [reply]
I have added an `echo` at the end of the prompt.

**Reply by faho** [reply]
> I found that you are doing a rewrite (in rust), if possible then please add this feature.

We try to be careful with adding configuration options, and I don't want to add this, so it is likely it won't be.

This is a case where fish and zsh differ quite a bit in philosophy - zsh is everything-optional-all-the-time, and fish very intentionally is not.

The rust port does not change anything about this.
